### PR TITLE
chore!: upgrade `fuel-core` to `0.24.3`

### PR DESCRIPTION
**Pull request by Torres-ssf**
- Closes: https://github.com/FuelLabs/fuels-ts/issues/1831
- Closes: https://github.com/FuelLabs/fuels-ts/issues/1700
- Closes: https://github.com/FuelLabs/fuels-ts/issues/1738

## Breaking Changes
- Requires `Forc` `0.51.1`
- Setting `maxFee` is always required now ( The SDK estimates it by default with `Provider.getTransactionCost` + `Account.fund` )
- `gasPrice` is calculated by the VM so we do not need to inform it anymore. The Policy `gasPrice` [was replaced](https://github.com/FuelLabs/fuels-ts/pull/1764/commits/3ad83744048ce4b4a28a6151048adcdf27aa9566) by `tip`
- `Account.fund` signature has [changed](https://github.com/FuelLabs/fuels-ts/pull/1764/commits/0851940f91620696fe4b481aac63fc3b4ad340ab)
- Graphql URL now includes a versioning path: `http://127.0.0.1:4000/v1/graphql` 
- `calculateTransactionFee` now requires [tip](https://github.com/FuelLabs/fuels-ts/pull/1764/commits/277ead08b124651224b4b8b1934c701b59471f3e), [maxGasPerTx](https://github.com/FuelLabs/fuels-ts/pull/1764/commits/eeb5b1333006253f8fbbb48509b62aca0c798f39), and [gasPrice](https://github.com/FuelLabs/fuels-ts/pull/1764/commits/499d78b887b8779fbd6913ecca710cc5fad3f83c)
- `calculateTransactionFee` does not receive [gasUsed](https://github.com/FuelLabs/fuels-ts/pull/1764/commits/499d78b887b8779fbd6913ecca710cc5fad3f83c) anymore
- [removed](https://github.com/FuelLabs/fuels-ts/pull/1764/commits/87bb4411ef1e1ec767c9fe811ab515240b854182) unneeded properties returned by `Provider.getTransactionCost`

Current Tests Passing:
Screenshot 2024-04-08 at 10 38 54

### CODE REVIEW

File: packages/account/src/providers/operations.graphql
Comment (nedsalk) [inline packages/account/src/providers/operations.graphql:19]: Adding receipts here requires updating the subscription logic to account for events that span multiple batches of streamed data because receipts can be arbitrarily large and can exceed the size of a single batch. I already have the logic for that implemented for #1644 but that branch is really old. I'll create a PR onto `master` with only that logic.

Comment (arboleya) [reply packages/account/src/providers/operations.graphql:19]: Sorry, but I lost context — what are you saying exactly?

Comment (nedsalk) [reply packages/account/src/providers/operations.graphql:19]: Sorry for the vagueness. After looking at the code again I realized I mixed up two things. I tagged you thinking that we'd need to push for the implementation of the [issue](https://github.com/FuelLabs/fuel-core/issues/1724) that's blocking #1644, but that's not the problem here. I'll close this discussion and write another comment detailing the issue.

Comment (arboleya) [reply packages/account/src/providers/operations.graphql:19]: Perfect, thank you.

File: packages/account/src/providers/message.ts
Comment (xgreenx) [inline packages/account/src/providers/message.ts:44]: This field was renamed to `message_outbox_root`, but still is available=) Also, we have 3 more fields:
- `event_inbox_root`
- `consensus_parameters_version`
- `state_transition_bytecode_version`

File: packages/account/src/providers/operations.graphql
Comment (nedsalk) [inline packages/account/src/providers/operations.graphql:36]: Because `transactionStatusFragment` is used in subscriptions as well, adding `receipts` directly onto it would mean that we'd be fetching them via `submitAndAwait`/`statusChange` subscriptions for every `Provider.sendTransaction` call and `TransactionResponse` creation even though we wouldn't be using them because we're taking the receipts from the response of `getTransactionWithReceipts`.

I suggest that you keep this `transactionStatusFragment` as you modified it, but create a separate `transactionStatusSubscriptionFragment` that'd only cover our current subscription needs. It'd look like this:

```gql
fragment transactionStatusSubscriptionFragment on TransactionStatus {
  type: __typename
  ... on SubmittedStatus {}
  ... on SuccessStatus {}
  ... on FailureStatus {}
  ... on SqueezedOutStatus {
    reason
  }
}

subscription submitAndAwait($encodedTransaction: HexString!) {
  submitAndAwait(tx: $encodedTransaction) {
    ...transactionStatusSubscriptionFragment
  }
}

subscription statusChange($transactionId: TransactionId!) {
  statusChange(id: $transactionId) {
    ...transactionStatusSubscriptionFragment
  }
}
```

I didn't try it, but maybe you could even get away with this fragment:

```gql
fragment transactionStatusSubscriptionFragment on TransactionStatus {
  type: __typename
  ... on SqueezedOutStatus {
    reason
  }
}
```


#1644 will use the receipts from `TransactionStatus` and reevaluate how receipts are integrated into `TransactionResponse`, etc.

Comment (Torres-ssf) [reply packages/account/src/providers/operations.graphql:36]: @nedsalk I will take a look into this

File: packages/account/src/providers/operations.graphql
Comment (nedsalk) [reply packages/account/src/providers/operations.graphql:19]: https://github.com/FuelLabs/fuels-ts/pull/2021#discussion_r1560930082

File: packages/account/src/providers/message.ts
Comment (Torres-ssf) [reply packages/account/src/providers/message.ts:44]: @xgreenx I've missed that.

Thanks a lot 🙏

File: packages/account/src/providers/operations.graphql
Comment (Torres-ssf) [reply packages/account/src/providers/operations.graphql:36]: @nedsalk I've added the receipts to the statuses fragments because starting on `0.23` receipts were removed from the root of the Transaction type on the graphQL schema and are now only present within the statuses.

Comment (nedsalk) [reply packages/account/src/providers/operations.graphql:36]: Yes, I understood that and that's why I said 
>[...] I suggest that you keep this transactionStatusFragment as you modified it...

My comment relates to keeping the fragments used in the subscriptions the same or trimming them down to only the fields used with the subscriptions in our code base currently.

File: packages/forc/VERSION
Comment (arboleya) [inline packages/forc/VERSION:1]: It'd be good to check if this works with `forc` @  `0.53.0`:
 - https://github.com/FuelLabs/fuels-ts/pull/2045

This version includes many performance improvements and bug fixes regarding the new encoding.

Keep in mind that, speaking of the new encoding, we will still need this before the end:
 - https://github.com/FuelLabs/fuels-ts/pull/2044

File: packages/fuel-gauge/src/predicate/predicate-with-script.test.ts
Comment (Torres-ssf) [reply packages/fuel-gauge/src/predicate/predicate-with-script.test.ts:66]: Mostly because of this: https://github.com/FuelLabs/fuels-ts/issues/2081

We have predicates being used by multiple test cases and since the tests are running concurrently, every type of verification that tries to predict a predicate balance will be flaky.

But `toBeGreaterThan` it is less flaky than `toEqual`

File: packages/fuel-gauge/src/predicate/predicate-with-contract.test.ts
Comment (Torres-ssf) [reply packages/fuel-gauge/src/predicate/predicate-with-contract.test.ts:118]: Mostly because of this: https://github.com/FuelLabs/fuels-ts/issues/2081

We have predicates being used by multiple test cases and since the tests are running concurrently, every type of verification that tries to predict a predicate balance will be flaky.

But `toBeGreaterThan` it is less flaky than `toEqual`

File: packages/account/src/wallet/base-wallet-unlocked.ts
Comment (Torres-ssf) [reply packages/account/src/wallet/base-wallet-unlocked.ts:111]: @petertonysmith94 Yes.

The `estimateTxDependecies` method is already called by `getTransactionCost`.

This means that we usually want this flag to be `false`.

File: packages/account/src/test-utils/launchNode.ts
Comment (Torres-ssf) [reply packages/account/src/test-utils/launchNode.ts:183]: Because now the state chain config JSON file accepts only numbers when defining UTXOs and message amounts.

This hack was required because all amounts need to be defined as U64 and in the number format. 

However JS does not support U64, and when the JSON file is read, JS messes up all almont. 

So I am defining the amounts as strings to avoid JS mutating the values. Later, after stringifying everything I use a regex to remove `'` from the string numbers.

I am using `any` so I can force all amounts to be a string since TS is not happy because, in the JSON file, they are numbers

File: packages/abi-coder/src/utils/constants.ts
Comment (Torres-ssf) [reply packages/abi-coder/src/utils/constants.ts:36]: https://github.com/FuelLabs/fuel-vm/pull/698/files

Previous to VM 0.48, UTXOs IDs had 33 bytes:
- The first 32 bytes is the Transaction ID that created the UTXO
- The following 1 byte (u8) is the output index where the `OutputCoin` that originated this UTXO was placed within the `outputs` of the Transaction.

On VM 0.48 the output index changed from u8 to u16, so now UTXO IDs have 34 bytes instead of 33.

File: packages/account/src/providers/transaction-request/transaction-request.ts
Comment (Torres-ssf) [reply packages/account/src/providers/transaction-request/transaction-request.ts:377]: It is already there: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/a168a52fd94d0c449d1aff37a5730c0ef7e97f34

File: packages/account/src/providers/transaction-request/transaction-request.ts
Comment (Torres-ssf) [reply packages/account/src/providers/transaction-request/transaction-request.ts:420]: It is already there: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/a168a52fd94d0c449d1aff37a5730c0ef7e97f34

File: packages/forc/VERSION
Comment (Torres-ssf) [reply packages/forc/VERSION:1]: @arboleya I've tried `0.55.0` and it is working fine

File: packages/fuel-gauge/src/script-with-vectors.test.ts
Comment (arboleya) [reply packages/fuel-gauge/src/script-with-vectors.test.ts:44]: This change was already applied in [#2045](https://github.com/FuelLabs/fuels-ts/pull/2045/files#diff-ddb8edd128f2fa94c012f00b28dc14b16100ff9aef89a03b7f1c609dda7d8c71).

File: packages/fuel-gauge/src/script-with-vectors.test.ts
Comment (arboleya) [reply packages/fuel-gauge/src/script-with-vectors.test.ts:50]: This change was already applied in https://github.com/FuelLabs/fuels-ts/pull/2045.

File: packages/account/src/providers/transaction-request/transaction-request.test.ts
Comment (Torres-ssf) [reply packages/account/src/providers/transaction-request/transaction-request.test.ts:80]: I've just increased the amount of fake UTXOs generated from the BaseAssetId.

This is being done to ensure the estimation dry-run will carry enough resources to cover the fee

File: packages/account/src/providers/transaction-request/input.ts
Comment (nedsalk) [inline packages/account/src/providers/transaction-request/input.ts:71]: typo: `padPredicateData`

File: packages/account/src/providers/transaction-request/transaction-request.ts
Comment (nedsalk) [inline packages/account/src/providers/transaction-request/transaction-request.ts:666]: From what I understand, `shiftPredicateData` and `padPredicateData` stem from us setting the predicate data "in stone" _before_ we encode the actual transaction into bytes and this is why both of these functions feel like a workaround. `Predicate.populateTransactionPredicateData` could fall into this category as well. IMO we should definitely refactor the whole transaction encoding workflow so that the dynamic elements of a transaction like `predicateData` get set only when `transaction.encodeTransaction()` is called. This would be another issue, of course; I wanted to get it out here so that it's on our radar.

Comment (Torres-ssf) [reply packages/account/src/providers/transaction-request/transaction-request.ts:666]: That is not possible right now. 

The `predicateData` needs to be in place when we estimate predicate gas used. It must also be in place when estimating gas and fees before even running the dry-run to get the `gasUsed`.

We need the `predicateData` before even funding the transaction.

Comment (nedsalk) [reply packages/account/src/providers/transaction-request/transaction-request.ts:666]: > That is not possible right now.

Not possible because of the scope of the task or technically not possible? If it's because of the scope of the task, I agree. I don't see technical limitations for it, though.

> The predicateData needs to be in place when we estimate predicate gas used. It must also be in place when estimating gas and fees before even running the dry-run to get the gasUsed.
We need the predicateData before even funding the transaction.

Every transaction has to be converted to bytes at one point before being sent to the node, be it for full execution, for dry-running, or whatever else (e.g. [tx estimation](https://github.com/FuelLabs/fuels-ts/blob/8f62b941193683319181d76a12c7e04effc2e568/packages/account/src/providers/provider.ts#L737)) My comment is  related to this tx-to-bytes conversion: that we resolve all these dynamic aspects of a transaction inside of `TransactionRequest.toTransactionBytes` - at the last moment, just before a transaction is sent.

File: packages/account/src/account.ts
Comment (nedsalk) [inline packages/account/src/account.ts:317]:
```suggestion
    const requestToReestimate = clone(txRequest);
    if (addedSignatures) {
      Array.from({ length: addedSignatures }).forEach(() =>
        requestToReestimate.addEmptyWitness()
      );
    }

    const { maxFee } = await this.provider.estimateTxGasAndFee({
      transactionRequest: requestToReestimate,
    });
```

Variable rename

File: packages/account/src/predicate/predicate.ts
Comment (nedsalk) [inline packages/account/src/predicate/predicate.ts:123]: We can remove this method from `Predicate` altogether now as it's doing nothing additional.

File: packages/account/src/providers/transaction-request/transaction-request.ts
Comment (nedsalk) [reply packages/account/src/providers/transaction-request/transaction-request.ts:666]: Created #2156 for this where we can continue the discussion.

File: packages/account/src/providers/operations.graphql
Comment (nedsalk) [inline packages/account/src/providers/operations.graphql:44]: Tooling doesn't allow me to make a suggestion here, but you can simplify this fragment:
```gql
fragment transactionStatusSubscriptionFragment on TransactionStatus {
  type: __typename
  ... on SqueezedOutStatus {
    reason
  }
}
```
The `type` and `SqueezedOutStatus.reason` fields are the only ones used so we can reduce bandwidth by removing the other ones. I tried it locally and tests passed.

File: packages/account/src/providers/provider.ts
Comment (nedsalk) [inline packages/account/src/providers/provider.ts:860]:
```suggestion
      for (let i = 0; i < dryRunResults.dryRun.length; i++) {
        const requestIdx = transactionsToProcess[i];
        const { receipts: rawReceipts, status } = dryRunResults.dryRun[i];
        const result = results[requestIdx];
        result.receipts = rawReceipts.map(processGqlReceipt);
        result.dryrunStatus = status;

        const { missingOutputVariables, missingOutputContractIds } = getReceiptsWithMissingData(
          result.receipts
        );
        const hasMissingOutputs =
          missingOutputVariables.length > 0 || missingOutputContractIds.length > 0;

        const request = allRequests[requestIdx];

        if (hasMissingOutputs && request?.type === TransactionType.Script) {
          result.outputVariables += missingOutputVariables.length;
          request.addVariableOutputs(missingOutputVariables.length);
          missingOutputContractIds.forEach(({ contractId }) => {
            request.addContractInputAndOutput(Address.fromString(contractId));
            result.missingContractIds.push(contractId);
          });

          const { maxFee } = await this.estimateTxGasAndFee({
            transactionRequest: request,
          });
          request.maxFee = maxFee;

          // Prepare for the next round of dry run
          serializedTransactionsMap.set(requestIdx, hexlify(request.toTransactionBytes()));
          nextRoundTransactions.push(requestIdx);
        }
      }
```

A bit of variable renaming and the like I did while trying to understand what's happening.

File: packages/account/src/providers/transaction-request/transaction-request.ts
Comment (nedsalk) [inline packages/account/src/providers/transaction-request/transaction-request.ts:467]: Why is this being removed here? It's being removed in #2037 as well.

File: packages/fuel-gauge/src/policies.test.ts
Comment (Torres-ssf) [reply packages/fuel-gauge/src/policies.test.ts:256]: I've re-added https://github.com/FuelLabs/fuels-ts/pull/2021/commits/363aff1cb486c85cb6dd590cffc62f20f3618c30

File: packages/fuel-gauge/src/policies.test.ts
Comment (Torres-ssf) [reply packages/fuel-gauge/src/policies.test.ts:55]: I've re-added https://github.com/FuelLabs/fuels-ts/pull/2021/commits/363aff1cb486c85cb6dd590cffc62f20f3618c30

File: packages/account/src/providers/transaction-request/input.ts
Comment (Torres-ssf) [reply packages/account/src/providers/transaction-request/input.ts]: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/3b5f0cad42ae92d4e3dffba17c9de104cb0daf78

File: packages/account/src/providers/transaction-request/input.ts
Comment (Torres-ssf) [reply packages/account/src/providers/transaction-request/input.ts]: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/3b5f0cad42ae92d4e3dffba17c9de104cb0daf78

File: packages/account/src/providers/message.ts
Comment (Torres-ssf) [reply packages/account/src/providers/message.ts]: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/3b5f0cad42ae92d4e3dffba17c9de104cb0daf78

File: packages/account/src/providers/operations.graphql
Comment (Torres-ssf) [reply packages/account/src/providers/operations.graphql:44]: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/a3e9c6e369d883bf311580014413deedbb757626

File: packages/account/src/account.ts
Comment (Torres-ssf) [reply packages/account/src/account.ts]: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/3f4e56f8f7dbdb6586670df08b649b8f9db5d190

File: packages/account/src/providers/provider.ts
Comment (Torres-ssf) [reply packages/account/src/providers/provider.ts]: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/2a91c098b42b513fb4459f1ff9367ead1545201a

File: packages/account/src/account.ts
Comment (Torres-ssf) [reply packages/account/src/account.ts]: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/be977efd1074a45254d04d60808a1e82e673fbbf

File: packages/account/src/providers/transaction-request/transaction-request.ts
Comment (Torres-ssf) [reply packages/account/src/providers/transaction-request/transaction-request.ts:467]: @nedsalk Yeah I've removed it first here.

It would be good if we could merge https://github.com/FuelLabs/fuels-ts/pull/2037 before this one.

I'm just waiting for one more review since it is a tricky change

File: packages/account/src/predicate/predicate.ts
Comment (Torres-ssf) [reply packages/account/src/predicate/predicate.ts]: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/6cc858b577fcef20351267211d2ee1c3017522f3

File: packages/fuel-gauge/src/predicate/predicate-with-script.test.ts
Comment (Torres-ssf) [reply packages/fuel-gauge/src/predicate/predicate-with-script.test.ts:69]: I reckon we don't need that mate. 

Many of these balance validations were added only as additional validations, and most do not add value to the test context. I've added many of them. 

Not that our tests are running concurrently, these validation become problematic since we have multiple tests using our predicates

**Comment by Torres-ssf** [comment]
> Hey @Torres-ssf, I have strong admiration for your work 🚀

@petertonysmith94 Thanks mate ❤️

> * Why change the amounts? Is it due to the changes to the chain config? [Example](https://github.com/FuelLabs/fuels-ts/pull/2021/files#diff-76ce4909e9c50e03eb8b42635b678afbb3f154b4fd823f336bd41d1d4d93382fR988-R990).

I've updated many funding amounts in our tests because transaction fees went considerably higher with these upgrades

> * Missing changeset for `0.24.2`

Added: https://github.com/FuelLabs/fuels-ts/pull/2021/commits/6006bce290740f85a287aa7bd550590be40cca28

> * Is it worth adding references to `// TODO: remove skip from testnet test` related tests in [Update old / beta5 URLs #2111](https://github.com/FuelLabs/fuels-ts/issues/2111)?

I will create an issue for it.

**Review by arboleya — APPROVED** [review]

**Review by arboleya — DISMISSED** [review]

**Review by arboleya — APPROVED** [review]

**Review by maschad — APPROVED** [review]

File: packages/account/src/test-utils/launchNode.ts
Comment (arboleya) [reply packages/account/src/test-utils/launchNode.ts:1]: Here he comes. 🕵️‍♂️ 🚨

**Pull request merged**
